Fix #10416  Do not copy ee9 response headers

gregw opened the pull request:
Use the core response HttpFields directly as the ee9 response headers to avoid copy and retain persistent field behaviour.

gregw committed:
Fix #10416 EE9 Response headers

Use the core response HttpFields directly as the ee9 response headers to avoid copy and retain persistent field behaviour.

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/12/10416/ee9-response-headers

janbartel reviewed (approved):
Seems fine. Is it worthwhile having a  test for setting the "Server" header, which is what revealed this problem in the first place?

gregw committed:
Fix #10416 EE9 Response headers

Use the core response HttpFields directly as the ee9 response headers to avoid copy and retain persistent field behaviour.

Added EE9 test to show that Persistent fields can be modified

Updated fix for #10339 so that persistent fields revert to original values after a clear operation

gregw commented:
@janbartel can you re-review... @lachlan-roberts can you review.
I've changed the fix for #10339 now so that when response fields are cleared they reset to their original values rather than keep their modified values.

lachlan-roberts commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/Response.java:
```diff
@@ -1272,25 +1277,7 @@ public void setTrailerFields(Supplier<Map<String, String>> trailers)
-        if (meta == null)
-            return newResponseMetaData();
-        return meta;
+        return new MetaData.Response(getStatus(), getReason(), _channel.getRequest().getHttpVersion(), _channel.getCoreResponse().getHeaders(), getLongContentLength(), getTrailers());
```
Why are you using `_channel.getCoreResponse().getHeaders()` and not `_fields`?

gregw replied:
They should now be the same instance.... but if somehow they are not the same, then we should use the ones form the core request.  I guess I could put in:
```suggestion
        assert _channel.getCoreResponse().getHeaders() == _fields
        return new MetaData.Response(getStatus(), getReason(), _channel.getRequest().getHttpVersion(), _channel.getCoreResponse().getHeaders(), getLongContentLength(), getTrailers());
```
But currently there is no way for that to not be true, so I'm good leaving it out unless you think it makes it clearer.

lachlan-roberts replied:
Yeah it just looks weird that all the modifications are done on the `_fields` `HttpFields.Mutable` which is passed in. But then when it is used we just rely on an external assumption that it is the one from the `getCoreResponse()`.

Do you even need `protected void onResponse(HttpFields.Mutable headers)` method when you could just assign `_fields = _channel.getCoreResponse().getHeaders()`?

lorban reviewed (approved):
LGTM
